crates/astria-sequencer/src/sequencer.rs

```diff
@@ -195,12 +236,31 @@ fn start_grpc_server(
         .add_service(ClientQueryServer::new(ibc.clone()))
         .add_service(ChannelQueryServer::new(ibc.clone()))
         .add_service(ConnectionQueryServer::new(ibc.clone()))
-        .add_service(SequencerServiceServer::new(sequencer_api));
+        .add_service(SequencerServiceServer::new(sequencer_api))
+        .add_optional_service(optimistic_block_service_server);
+
+    // let mut optimistic_block_inner_handle = Fuse::terminated();
+    let optimistic_block_inner_handle = {
+        if no_optimistic_blocks {
+            Fuse::terminated()
+        } else {
+            tokio::task::spawn(async move { optimistic_block_service_inner.run().await }).fuse()
+        }
+    };
 
     info!(grpc_addr = grpc_addr.to_string(), "starting grpc server");
-    tokio::task::spawn(
-        grpc_server.serve_with_shutdown(grpc_addr, shutdown_rx.unwrap_or_else(|_| ())),
-    )
+
+    tokio::task::spawn(grpc_server.serve_with_shutdown(grpc_addr, async move {
+        tokio::select! {
+            _ = shutdown_rx => {
+                info!("grpc server shutting down");
+            },
+            _ = optimistic_block_inner_handle => {
+                info!("optimistic block inner handle task exited");
+            }
+        }
+        optimistic_streams_cancellation_token.cancel();
```

Comment:
After this, wait a moment so that the service has time to shutdown (only necessary if you didn't get here because of the task optimistic stream task having shut down).